[internal/exp/metrics] Add a new internal package for handling metric staleness

RichieSams opened the pull request:
**Description:**
It's a glorified wrapper over a Map type, which allows values to be expired based on a pre-supplied interval.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31016

**Testing:**
I added some basic tests of the PriorityQueue implementation as well as the expiry behaviour of Staleness

**Documentation:**

All the new structs are documented

RichieSams committed:
[internal/exp/metrics] Add a new internal package for handling metric staleness

It's a glorified wrapper over a Map type, which allows values to be expired based on a pre-supplied interval.

sh0rez commented on internal/exp/metrics/staleness/staleness.go:
```diff
@@ -0,0 +1,55 @@
+	Delete(key identity.Stream)
+	// Range calls f sequentially for each key and value present in the map.
+	// If f returns false, range stops the iteration.
+	Range(f func(key identity.Stream, value T) bool)
```
I think here is an opportunity to already conform to the future rangefunc api: https://go.dev/wiki/RangefuncExperiment

```go
func (m RawMap[T]) Items() iter.Seq2[identity.Stream, T] {
    return func(yield func(identity.Stream, T) bool) bool {
        for k, v := range m {
            if !yield(k, v) { break }
        }
        return false
    }
}
```

RichieSams replied:
That said, it looks like for `iter.Seq2` to be defined, we need go 1.22 and `GOEXPERIMENT=rangefunc`. Which I don't think we can allow for all the contrib.

Should we implement that ourselves for now? Or stick with Range until iter.Seq2 is GA?

sh0rez replied:
iter.Seq2 is merely an alias for `func(yield func(K, V) bool) bool`, so we can also just this raw type in the signature, which will be forward compatible!

RichieSams replied:
Ah good point. Thanks!

sh0rez commented on internal/exp/metrics/staleness/staleness_test.go:
```diff
@@ -0,0 +1,121 @@
+type RawMap[K comparable, V any] map[K]V
+
+func (rm *RawMap[K, V]) Load(key K) (V, bool) {
+	value, ok := (*rm)[key]
```
given the cache-like nature, I think callers would benefit from having `Load()` initialize to zero-value when not-exist, eliminating the need for a load-check-store

RichieSams replied:
We could add `LoadOrStore(key K, value V) (V, bool)`. Which would do what you're saying.

RichieSams replied:
Added

sh0rez commented on internal/exp/metrics/staleness/priority_queue.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+func (pq *PriorityQueue) Pop() (identity.Stream, time.Time) {
+	val := heap.Pop(&pq.inner).(*queueItem)
```
does pop also have to delete the item from the lookup map, so it doesn't hold onto every historical item pointer indefinitely?

RichieSams replied:
Fixed

sh0rez commented on internal/exp/metrics/staleness/priority_queue.go:
```diff
@@ -0,0 +1,97 @@
+	heap.Fix(pq, item.index)
+}
+
+type PriorityQueue struct {
```
wdyt about `PriorityQueue` (as used in the Staleness type) being an interface and having something like private like `heapQueue` be the actual implementation. Couples those two a little less

RichieSams replied:
Done

tombrk reviewed (changes requested):
This is really nice and exactly what I had in mind. Great work!!

RichieSams committed:
Properly clean up lookup map when popping

RichieSams committed:
Make PriorityMap an interface

So Staleness isn't so tightly tied to it

RichieSams committed:
Implement Map.LoadOrStore()

RichieSams committed:
Use the new upcoming iterator style instead of our custom Range()

sh0rez commented on internal/exp/metrics/staleness/staleness.go:
```diff
@@ -0,0 +1,61 @@
+	Store(key identity.Stream, value T)
+	// LoadOrStore will either load the value from the map and return it and the boolean `true`
+	// or if it doesn't exist in the Map yet, the value passed in will be stored and then returned with the boolean `false`
+	LoadOrStore(key identity.Stream, value T) (T, bool)
+	// Remove the value at key from the map
```
would like to revert my previous opinion, this is actually a lot less necessary than I thought.
lets remove it to keep the api surface smaller ;)

tombrk reviewed (commented):
tried this out, works really well!
I think this is ready for final review once rebased onto HEAD!

RichieSams committed:
Move Staleness map to be internal

So users are forced to use the correct methods.

Also adds lots of documentation

RichieSams committed:
Fix linter / check errors

RichieSams commented:
@jpkrohling @fatsheep9146 @bryan-aguilar 

This PR is ready for final review. (Can one of you add the "Skip Changelog" label, since this is an internal-only change)

@sh0rez Any final comments?

sh0rez commented on internal/exp/metrics/staleness/map.go:
```diff
@@ -0,0 +1,54 @@
+var _ Map[time.Time] = (*RawMap[identity.Stream, time.Time])(nil)
+
+// RawMap is an implementation of the Map interface using a standard golang map
+type RawMap[K comparable, V any] map[K]V
```
I think this type doesn't creeps a bit beyond the scope of this package.
I like that this package defines the `Map[T]` interface as the behavior it expects to fulfill its purpose.
A general stream map would feel more fitting as a `streams.HashMap` or similar.

I'd be okay with this pkg not defining a map implementation at all, leaving that to consumers.
No hard opinion however, if you prefer we can leave it in here for now, as we'll likely iterate on these packages for a bit anyways.

RichieSams replied:
> I'd be okay with this pkg not defining a map implementation at all, leaving that to consumers.

That was my initial plan. I had implemented RawMap just in the tests. But then I realized that I would need to re-implement the same code in the interval processor (and perhaps you as well in deltatocumulative). So I moved it to be in the package proper.

But I'm kind of the same. I don't feel strongly either way

RichieSams replied:
Are you ok with leaving it in for now? And iterating as the various processors utilize it?

RichieSams replied:
Bump

tombrk reviewed (commented):
I think it's in great shape for the initial bring-up. Everything else can be iterated upon :)

djaglowski reviewed (commented):
LGTM but I will wait to approve until @sh0rez's feedback is fully addressed.

RichieSams committed:
The Staleness tests can't be parallel, because they use the global `NowFunc`

RichieSams committed:
Expose `NowFunc` so other modules can do testing

sh0rez commented:
@djaglowski ready to merge from my side!

Think this still needs the "no changelog" label
